Updated fix for Perl programs on Darwin

bdesham opened the pull request:
###### Motivation for this change

Fix biber (#35353) and exiftool (#63111) on Darwin.

This PR is just an updated version of PR #35477 by @boronine. That PR has merge conflicts now; this one fixes the conflicts. (@boronine: if you would rather pull these changes into your branch and rebase, please feel free to do so! I don’t want to seem as if I’m trying to take credit for your work.)

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [ ] NixOS
   - [x] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

--
Resolves: #35353 #35477  #47302 #62156 #63111

volth commented:
It seems overkill.
Why not to add in `biber` and `exiftool` 3 lines line
```perl
 preConfigure = lib.optionalString stdenv.isDarwin ''
   patch exact file you need to patch
 '';
```

bdesham commented:
I don’t think it’s just a matter of patching, though. On my system the shebang line for the `biber` built from the latest Nixpkgs master is almost 50 KB (while the maximum macOS will tolerate is 512 bytes). I’m not sure how we’d get around this without introducing a wrapper. Can you be more specific about how we could just patch biber to fix it on macOS? `patchShebangs` doesn’t do anything to improve the length of the shebang line.

(The shebang line for exiftool contains a lot of redundant `-I` options, so it could be manipulated via `sed` to be less than 512 characters. But this approach is brittle and it wouldn’t work for biber anyway.)

volth commented:
> the shebang line for the biber built the latest Nixpkgs master is almost 50 KB (while the maximum macOS will tolerate is 512 bytes).

Does `perl.withPackages` work on Darwin?
it is a way to add packages to perl without adding `-I` to shebang lines

volth commented:
try this for `biber`:

```
{ stdenv, perl, perlPackages, texlive }:

let
  biberSource = stdenv.lib.head (builtins.filter (p: p.tlType == "source") texlive.biber.pkgs);
  deps = p: with p; [
    autovivification BusinessISBN BusinessISMN BusinessISSN ConfigAutoConf
    DataCompare DataDump DateSimple EncodeEUCJPASCII EncodeHanExtra EncodeJIS2K
    DateTime DateTimeFormatBuilder DateTimeCalendarJulian
    ExtUtilsLibBuilder FileSlurper FileWhich IPCRun3 LogLog4perl LWPProtocolHttps ListAllUtils
    ListMoreUtils MozillaCA ReadonlyXS RegexpCommon TextBibTeX
    UnicodeLineBreak URI XMLLibXMLSimple XMLLibXSLT XMLWriter
    ClassAccessor TextCSV TextCSV_XS TextRoman DataUniqid LinguaTranslit SortKey
    TestDifferences
    PerlIOutf8_strict
  ];
in

perlPackages.buildPerlModule {
  pname = "biber";
  inherit (biberSource) version;

  src = "${biberSource}/source/bibtex/biber/biblatex-biber.tar.gz";

  buildInputs = (deps perlPackages);

  postInstall = ''sed -E "s|^#\!.+$|#\!${perl.withPackages deps}/bin/perl -I$out/${perl.libPrefix}|" -i $out/bin/biber'';
  dontPatchShebangs = true;

  meta = with stdenv.lib; {
    description = "Backend for BibLaTeX";
    license = with licenses; [ artistic1 gpl1Plus ];
    platforms = platforms.unix;
    maintainers = [ maintainers.ttuegel ];
  };
}
```

volth commented:
For exiftool, its sheband line looks nice, the only problem is duplicated `-I` entries

 * as a quick hack, yes, `sed`:
```nix
{
 postInstall = ''sed -E "s|^#\!.+$|#\!${perl}/bin/perl -I$out/${perl.libPrefix}|" -i $out/bin/exiftool'';
 dontPatchShebangs = true;
}
```

 * as a better solution (which would require ~2 months way through `staging` because of mass-rebuild is deduplicate `-I` entries (then the `sed` hack could be removed)

volth commented:
@GrahamcOfBorg build biber

volth commented:
or
```
{ stdenv, perl, perlPackages, texlive, makeWrapper }:

let
  biberSource = stdenv.lib.head (builtins.filter (p: p.tlType == "source") texlive.biber.pkgs);
in

perlPackages.buildPerlModule {
  pname = "biber";
  inherit (biberSource) version;

  src = "${biberSource}/source/bibtex/biber/biblatex-biber.tar.gz";

  buildInputs = with perlPackages; [
    autovivification BusinessISBN BusinessISMN BusinessISSN ConfigAutoConf
    DataCompare DataDump DateSimple EncodeEUCJPASCII EncodeHanExtra EncodeJIS2K
    DateTime DateTimeFormatBuilder DateTimeCalendarJulian
    ExtUtilsLibBuilder FileSlurper FileWhich IPCRun3 LogLog4perl LWPProtocolHttps ListAllUtils
    ListMoreUtils MozillaCA ReadonlyXS RegexpCommon TextBibTeX
    UnicodeLineBreak URI XMLLibXMLSimple XMLLibXSLT XMLWriter
    ClassAccessor TextCSV TextCSV_XS TextRoman DataUniqid LinguaTranslit SortKey
    TestDifferences
    PerlIOutf8_strict
  ];

  nativeBuildInputs = [ makeWrapper ];

  perlPostHook = stdenv.lib.optionalString stdenv.isDarwin ''
    wrapProgram $out/bin/biber --set PERL5LIB $PERL5LIB
  '';

  meta = with stdenv.lib; {
    description = "Backend for BibLaTeX";
    license = with licenses; [ artistic1 gpl1Plus ];
    platforms = platforms.unix;
    maintainers = [ maintainers.ttuegel ];
  };
}
```

shebang like will remain huge, but it is not used as shebang, because `$out/bin/.biber-wrapped` is not executed as a shell script, so it should work

risicle commented:
Builds ok for me macos 10.13 FWIW

gpampara commented:
Although the build works, the same error is present - the error is now just in the resulting `.biber-wrapped` script. The wrapper script does correctly set the environment though.

bdesham commented:
@volth I still don’t understand why creating an ad-hoc solution for each different program is preferable to creating a single wrapper that can be invoked when needed. Is there some concern with the closure size?

bdesham commented:
@gpampara Are you having that issue with the code from the PR, or one of the versions from the comments here?

The PR version of Biber works for me on macOS 10.13, at least to the extent that `biber --help` and `biber --version` work fine. The shebang lines for both scripts are nice and short:

```
$ head -n 1 result/bin/biber result/bin/.biber-wrapped 
==> result/bin/biber <==
#! /nix/store/5frlp3yii9vhxx6sc4daw3vzs7r3w7k5-bash-4.4-p23/bin/bash -e

==> result/bin/.biber-wrapped <==
#!/nix/store/3mxzksvycp6sk730zqqzch3qwla3hfhd-perl-5.28.2/bin/perl
```

volth commented:
> @volth I still don’t understand why creating an ad-hoc solution for each different program is preferable to creating a single wrapper that can be invoked when needed.

Because there is only one such program - `biber` (`exiftool` will be fixed in https://github.com/NixOS/nixpkgs/pull/66458) and it has problem on Darwin.
Is it a reason to bloat `2x` the generic function used to build 1000+ other packages ?
Or should the fix be in `biber` fonder under `if stdenv.isDarwin` ?

bdesham commented:
@volth It seems like ack has the same problem on Darwin too—see #47302—and [this comment](https://github.com/NixOS/nixpkgs/pull/35477#issuecomment-455677546) implies that the youtube-viewer package is affected as well. There doesn’t seem to be a ticket for youtube-viewer but I went ahead and updated this PR to include a fix for ack too.

I’m not opposed to taking the ad-hoc approach if it really is only two packages that are affected, but if every Perl-based program has the potential to trigger this problem on Darwin then it seems like the fix should be part of the Nix Perl infrastructure, not something for each program to solve on its own. I understand that you don’t want to include a near-duplicate wrapper if it’s not necessary, but isn’t it going to be more total lines of code to fix the same problem in every single affected package?

gpampara commented:
@bdesham the patch listings from the comments. I was hoping that they would provide a quick solution, but I'll apply your patches on a pinned version of nixpkgs locally :)

volth commented:
> eally is only two packages that are affected, but if every Perl-based program has the potential to trigger this problem on Darwin then it seems like the fix should be part of the Nix Perl infrastructure,

The problem is: most of the derivations produced by `buildPerlPackage` are not "Perl-based programs". They are not programs at all, they are Perl libraries.
So if the Darwin problem become mass, it would make sense to make `buildPerlProgram` (or `buildDarwinPerlProgram`)

volth commented:
`ack`'s shebang is short, seems good (at least `ack-3.0.2`, the ticket is about `ack-2.x`)
`youtube-viewer`, yes, it is the second `biber`.

volth commented:
BTW there is a wrapper-less way to avoid shebangs
https://github.com/NixOS/nixpkgs/pull/55786/files
Could it be made the default for Darwin?

bdesham commented:
That seems reasonable to me. Just to be clear, you’re okay with adding a `builder_darwin.sh` script or something like that? Or should I try to put the logic (and OS checking) inline in the existing `builder.sh`?

bdesham commented:
I guess creating a new builder for Darwin would prevent a whole bunch of rebuilding on other platforms.

volth commented:
Or maybe just to create an independent bash function akin to `wrapProgram` which would do the stupid trasnformation:
```diff
- #!perl -Iaaaa -Ibbbbb
+ #!perl
+ use lib 'aaaa', 'bbbb';
```
and add it to `biber` and `youtube-viewer` derivations?

(I tried to enable it globally instead of adding -I to shebang, and it broke a lot; Perl has many packages which process or just read other Perl sources and they are sensitive to adding `use lib`; but it should be ok for end user programs like `biber` and `youtube-viewer`)

volth commented:
Another problem with the initial approach:
`doUseWrapper` triggers two unrelated things at a time:
1. it disables patching shebangs on `preConfigure` phase, which is performed on all source files, including build and test scripts
2. it adds `makeWrapper` on `postFixup` phase which is performed on `$out/bin/*`

It is just a coincidence that both things process `$out/bin/biber` and did not break build nor tests, so I am not sure that it is generic enough to help every program which suffers from long shebangs.
So to say it is ad-hoc for `biber` moved into `buildPerlPackage`.

bdesham commented:
OK, I’ll work on a new Bash script we can call for these kinds of programs.

bdesham committed:
add shortenPerlShebang function

This setup hook modifies a Perl script so that any "-I" flags in its shebang
line are rewritten into a "use lib ..." statement on the next line. This gets
around a limitation in Darwin, which will not properly handle a script whose
shebang line exceeds 511 characters.

bdesham committed:
biber: fix too-long shebang line on Darwin

Closes #35353.

bdesham committed:
ack: fix too-long shebang line on Darwin

Also remove gnused from the buildInputs: it's already provided to us in
this stage.

Closes #47302.

bdesham committed:
exiftool: fix too-long shebang line on Darwin

Closes #63111.

bdesham commented:
I rewrote this PR to introduce a shell function called `shortenPerlShebang`. This function is invoked in the `postInstall` phase for ack, biber, exiftool, and youtube-viewer when building for Darwin. If this approach looks good then I will

- mention this function in the manual,
- add some inline documentation of the function,
- add a check to ensure that the file being operated on actually does have a Perl shebang line, and
- print a warning message if the resulting shebang line is still longer than 512 characters.

volth commented on pkgs/top-level/perl-packages.nix:
```diff
@@ -89,22 +89,30 @@ let
+
     # use gnused so that the preCheck command passes
-    buildInputs = stdenv.lib.optional stdenv.isDarwin gnused;
+    buildInputs = stdenv.lib.optionals stdenv.isDarwin [ gnused shortenPerlShebang ];
```
`gnused` is not needed, it should be already there

bdesham replied:
Fixed.

volth commented on pkgs/top-level/perl-packages.nix:
```diff
@@ -18961,12 +18974,16 @@ let
       sha256 = "15xyrwv08fw8jmpydwzks26ipxnzliwddgyjcfqiaj0p7lwlhmx1";
     };
 
+    buildInputs = stdenv.lib.optionals stdenv.isDarwin [ gnused shortenPerlShebang ];
```
same here

bdesham replied:
Fixed.

bdesham committed:
youtube-viewer: fix too-long shebang line on Darwin

bdesham commented:
I’ve made the additional improvements I mentioned, so this should be good to go.

veprbl reviewed (changes requested):
How was this tested? I get a following problem:
```
[nix-shell:~/.cache/nix-review/pr-66446]$ cd nixpkgs

[nix-shell:~/.cache/nix-review/pr-66446/nixpkgs]$ NIX_PATH=nixpkgs=`pwd` nix-shell -p biber --pure --run biber
/private/var/folders/1f/2mj_ng1n2y3_wc59fgw3fn_r0000gn/T/nix-shell-18812-0/rc: line 1: /nix/store/2rdg8x1vbrk839kngcbffchnm9847h7m-perl5.28.2-biber-2.12/bin/biber: Permission denied

[nix-shell:~/.cache/nix-review/pr-66446/nixpkgs]$ ls -la /nix/store/2rdg8x1vbrk839kngcbffchnm9847h7m-perl5.28.2-biber-2.12/bin/biber
-r--r--r-- 1 user wheel 41877 Dec 31  1969 /nix/store/2rdg8x1vbrk839kngcbffchnm9847h7m-perl5.28.2-biber-2.12/bin/biber
```
Seems like mv inherits permissions from the temp file.

veprbl commented:
This could be also useful for latexindent: https://github.com/NixOS/nixpkgs/pull/59537#issuecomment-491859817

bdesham commented:
@veprbl I did test the execution of the scripts at *some* point, but apparently I made other changes in the meantime that broke the files’ modes. Sorry about that! The issue should be fixed now.

veprbl commented on pkgs/top-level/perl-packages.nix:
```diff
@@ -89,22 +89,29 @@ let
+
+    buildInputs = stdenv.lib.optional stdenv.isDarwin shortenPerlShebang;
     propagatedBuildInputs = [ FileNext ];
+    postInstall = stdenv.lib.optional stdenv.isDarwin ''
```
```suggestion
    postInstall = stdenv.lib.optionalString stdenv.isDarwin ''
```

veprbl commented on pkgs/top-level/perl-packages.nix:
```diff
@@ -8855,6 +8862,11 @@ let
     };
 
+    buildInputs = stdenv.lib.optional stdenv.isDarwin shortenPerlShebang;
+    postInstall = stdenv.lib.optional stdenv.isDarwin ''
```
```suggestion
    postInstall = stdenv.lib.optionalString stdenv.isDarwin ''
```

veprbl commented on pkgs/tools/typesetting/biber/default.nix:
```diff
@@ -20,7 +20,11 @@ perlPackages.buildPerlModule {
-  ];
+  ] ++ stdenv.lib.optional stdenv.isDarwin shortenPerlShebang;
+
+  postInstall = stdenv.lib.optional stdenv.isDarwin ''
```
```suggestion
  postInstall = stdenv.lib.optionalString stdenv.isDarwin ''
```

veprbl commented on pkgs/top-level/perl-packages.nix:
```diff
@@ -19282,12 +19294,16 @@ let
       DataDump
       JSON
     ];
+    postInstall = stdenv.lib.optional stdenv.isDarwin ''
```
```suggestion
    postInstall = stdenv.lib.optionalString stdenv.isDarwin ''
```

veprbl commented:
Also should use `nativeBuildInputs` instead of `buildInputs`.

veprbl committed:
texlive: fix too-long shebang line on Darwin for latexinent

veprbl commented:
Thanks @bdesham @boronine @volth and others for tackling this bug!
